### PR TITLE
refactor: Refactoring to match the rest of terraform-aws-modules

### LINKED ISSUE #1245: Unable to have more than one profile in fargate_profiles
# I have issues

I'm using Terraform v0.14.6. When I try to add a new item to `fargate_profiles` I get an error during `terraform plan`.

## I'm submitting a...

* [X] bug report
* [ ] feature request
* [ ] support request - read the [FAQ](https://github.com/terraform-aws-modules/terraform-aws-eks/blob/master/docs/faq.md) first!
* [ ] kudos, thank you, warm fuzzy

## What is the current behavior?

`terraform plan` shows this:

```
Error: Inconsistent conditional result types

  on .terraform/modules/my_eks/modules/fargate/fargate.tf line 17, in resource "aws_eks_fargate_profile" "this":
  17:   for_each               = local.create_eks ? local.fargate_profiles_expanded : {}
    |----------------
    | local.create_eks is true
    | local.fargate_profiles_expanded is object with 4 attributes

The true and false result expressions must have consistent types. The given
expressions are object and object, respectively.
```

## If this is a bug, how to reproduce? Please include a code sample if relevant.

Have an instance of this module with `fargate_profiles`, after a successful `terraform apply` try to add a new item to `fargate_profiles` and run `terraform plan`.

EDIT: Looks like this happens whenever you have more than one fargate profile, not only when adding a new one (see comments below).

## What's the expected behavior?

`terraform plan` should work and show the new `aws_eks_fargate_profile` resources being created.

## Are you able to fix this problem and submit a PR? Link here if you have already.

I don't have a PR because I'm not sure what is correct solution for this, but I made a copy of this module locally and changed this:

```
resource "aws_eks_fargate_profile" "this" {
  for_each               = local.create_eks ? local.fargate_profiles_expanded : {}
```

With this:

```
resource "aws_eks_fargate_profile" "this" {
  for_each               = local.create_eks ? local.fargate_profiles_expanded : null
```

To workaround the issue for now.

In other words, I changed `{}` with `null`. Not even sure how `for_each` works with `null`, but since it didn't complain, I decided to leave it like this until we find the correct fix.

## Environment details

* Affected module version: 14.0.0
* OS: Some recent version of Amazon Linux I believe
* Terraform version: 0.14.6

Thanks!

### PR DESCRIPTION
**Pull request by antonbabenko**
This refactoring had an intention to keep all existing functionality and fix obvious issues:
1. critical bugs (like syntax errors)
2. remove unused variables and illogical parts of the code which were plain wrong
3. add documentation for the missing parts (variables and outputs)
4. update examples, so we can start trusting them again
5. add `tflint` hook and make it happy (without breaking anything)

Backward compatibility is important here.

Fixes #1532
Fixes #1245

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -29,33 +38,32 @@ By default, this module manages the `aws-auth` configmap for you (`manage_aws_au
...
 }
 
 provider "kubernetes" {
-  host                   = data.aws_eks_cluster.cluster.endpoint
-  cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
-  token                  = data.aws_eks_cluster_auth.cluster.token
+  host                   = data.aws_eks_cluster.eks.endpoint
+  cluster_ca_certificate = base64decode(data.aws_eks_cluster.eks.certificate_authority[0].data)
+  token                  = data.aws_eks_cluster_auth.eks.token
   load_config_file       = false
```
Comment (daroga0002) [inline README.md:57]: you must remove this as was deprecated in kubernetes v2 provider

File: examples/_bootstrap/main.tf [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+module "vpc" {
+  source  = "terraform-aws-modules/vpc/aws"
+  version = "~> 3.0"
+
+  name = local.name
+  cidr = "10.0.0.0/16"
+
+  azs             = data.aws_availability_zones.available.names
+  public_subnets  = [for k, v in data.aws_availability_zones.available.names : cidrsubnet(local.vpc_cidr, 8, k)]
```
Comment (daroga0002) [inline examples/_bootstrap/main.tf:32]: maybe more clear will be move this to locals?

File: examples/fargate/main.tf [hunk 3]
```diff
@@ -16,7 +12,7 @@ data "aws_eks_cluster_auth" "cluster" {
 
 provider "kubernetes" {
   host                   = data.aws_eks_cluster.cluster.endpoint
-  cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority.0.data)
+  cluster_ca_certificate = base64decode(data.aws_eks_cluster.cluster.certificate_authority[0].data)
   token                  = data.aws_eks_cluster_auth.cluster.token
   load_config_file       = false
```
Comment (daroga0002) [inline examples/fargate/main.tf:17]: not supported in v2 kubernetes provider

File: locals.tf [hunk 4]
```diff
@@ -1,53 +1,41 @@
...
+  cluster_id              = coalescelist(aws_eks_cluster.this[*].id, [""])[0]
+  cluster_arn             = coalescelist(aws_eks_cluster.this[*].arn, [""])[0]
+  cluster_name            = coalescelist(aws_eks_cluster.this[*].name, [""])[0]
+  cluster_endpoint        = coalescelist(aws_eks_cluster.this[*].endpoint, [""])[0]
+  cluster_auth_base64     = coalescelist(aws_eks_cluster.this[*].certificate_authority[0].data, [""])[0]
+  cluster_oidc_issuer_url = flatten(concat(aws_eks_cluster.this[*].identity[*].oidc[0].issuer, [""]))[0]
+
   cluster_security_group_id         = var.cluster_create_security_group ? join("", aws_security_group.cluster.*.id) : var.cluster_security_group_id
-  cluster_primary_security_group_id = var.cluster_version >= 1.14 ? element(concat(aws_eks_cluster.this[*].vpc_config[0].cluster_security_group_id, [""]), 0) : null
+  cluster_primary_security_group_id = try(var.cluster_version >= 1.14, false) ? element(concat(aws_eks_cluster.this[*].vpc_config[0].cluster_security_group_id, [""]), 0) : null
```
Comment (daroga0002) [inline locals.tf:12]: 1.14 can be removed at all as this is not supported anymore and there is no way to create it. 

Also AWS states:
```
Q: What happens on the end of support date?

A: On the end of support date, you can no longer create new Amazon EKS clusters with the unsupported version. Existing control planes are automatically updated by Amazon EKS to the oldest supported version through a gradual deployment process after the end of support date. After the automatic control plane update, you must manually update cluster add-ons and Amazon EC2 nodes. For more information, see To update the Kubernetes version for your Amazon EKS cluster .
```

https://docs.aws.amazon.com/eks/latest/userguide/kubernetes-versions.html

File: locals.tf [hunk 5]
```diff
@@ -1,53 +1,41 @@
...
-      null
-    )
-  ) == "windows"]) > 0
-
-  worker_ami_name_filter = var.worker_ami_name_filter != "" ? var.worker_ami_name_filter : "amazon-eks-node-${var.cluster_version}-v*"
+  worker_ami_name_filter = var.worker_ami_name_filter != "" ? var.worker_ami_name_filter : "amazon-eks-node-${var.cluster_version != null ? var.cluster_version : ""}-v*"
   # Windows nodes are available from k8s 1.14. If cluster version is less than 1.14, fix ami filter to some constant to not fail on 'terraform plan'.
   worker_ami_name_filter_windows = (var.worker_ami_name_filter_windows != "" ?
-    var.worker_ami_name_filter_windows : "Windows_Server-2019-English-Core-EKS_Optimized-${tonumber(var.cluster_version) >= 1.14 ? var.cluster_version : 1.14}-*"
+    var.worker_ami_name_filter_windows : "Windows_Server-2019-English-Core-EKS_Optimized-${try((tonumber(var.cluster_version) >= 1.14 ? var.cluster_version : 1.14), 1.14)}-*"
```
Comment (daroga0002) [inline locals.tf:31]: same a above, 1.14 is not existing anymore

**Review by daroga0002 — COMMENTED** [review]
I know it is still draft, but I taken a fast look what is going on and left some comments which should be helpful

**Comment by antonbabenko** [comment]
@daroga0002 Thanks for the initial review! Some of the comments are not relevant because I have not finished many of examples (WIP still). Let me ping you when I am ready for review - tomorrow or on Wednesday. I am changing a lot of files but trying to keep the same logic in all places.

File: locals.tf (hunk 4 above)
Comment (antonbabenko) [reply locals.tf:12]: Good point. I will add `minimum_cluster_version` variable with the default value "1.16".

File: README.md (hunk 1 above)
Comment (antonbabenko) [reply README.md]: done

File: examples/_bootstrap/main.tf (hunk 2 above)
Comment (antonbabenko) [reply examples/_bootstrap/main.tf:33]: No need because this is used just once.

File: locals.tf (hunk 4 above)
Comment (antonbabenko) [reply locals.tf]: Simplified to remove the check completely.

**Comment by antonbabenko** [comment]
@daroga0002 Please review this one.

The remaining bits in this PR are related to the examples (instance_refresh, irsa, launch_templates, launch_templates_with_managed_node_groups, managed_node_groups, secrets_encryption). I will look into that tomorrow morning.

**Review by daroga0002 — COMMENTED** [review]
I have mixed feelings about new example format where we need to execute a bootstrap (especially including a EKS there) because now when I will want to test soemthing then I must run bootstrap (15 minutes) and then second example (again 15 minutes).

Also dont like here that I must run one dir and then second (so I cannot just run it in background and return when it will be created).

One more point against it that potentially running few examples (I used to make such experiments) will create me kubernetes inside same VPC (created in bootstrap) what can cause some interference between clusters.

I am just loudly thinking but maybe better will be just exclude dependencies (VPC) into separate file  `vpc.tf` which will be just copied into each example directory?

Beside this it looks good 👍

**Comment by antonbabenko** [comment]
Thanks for the review and feedback!

I kind of agree that separation in examples is not very logical. At first, `bootstrap` didn't have EKS but just VPC which can be created once and it is free to keep running. Later, I worked on `modules/fargate` and respective examples for it which require the existing EKS cluster to put Fargate profiles into when the submodule is used separately.

I will change it now and move the EKS cluster from bootstrap into individual examples where it makes sense. This way we can run `terraform apply` just once in a single example folder we want to verify.

**Comment by antonbabenko** [comment]
[v17.19.0](https://github.com/terraform-aws-modules/terraform-aws-eks/releases/v17.19.0) has been just released.

If this release breaks something that was working before, let's fix it immediately. I did my best to not break anything and put breaking changes in the [milestone v18.0.0](https://github.com/terraform-aws-modules/terraform-aws-eks/milestone/5).

Let's make sure that new PRs are rebased properly before we do a review.

**Pull request merged**